src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1473,6 +1507,29 @@ public void ProcessFileChange(string projectSystemFilePath, string workspaceFile
                 return solution;
             }
 
+            internal Solution UpdateProjectDocumentsOrderForBatch(Solution oldSolution, ProjectId projectId)
+            {
+                var orderedFilePaths = _orderedFilePathsInBatch.ToImmutableArray();
+                if (orderedFilePaths.IsEmpty)
+                {
+                    return oldSolution;
+                }
+
+                _orderedFilePathsInBatch.Clear();
+
+                return UpdateProjectDocumentsOrder(oldSolution, projectId, orderedFilePaths);
+            }
+
+            private Solution UpdateProjectDocumentsOrder(Solution oldSolution, ProjectId projectId, IEnumerable<string> filePaths)
+            {
+                var documentIds =
+                    filePaths
+                        .Select(x => oldSolution.GetDocumentIdsWithFilePath(x).FirstOrDefault(id => id.ProjectId == projectId))
```

Comment:
Shouldn't this be `.Single`?  If a host is asking to order with this ordering, and the file they're asking about isn't even in the solution... that seems *really really bad*.

Reply:
Yes it does seem bad if they pass a file path that isn't in the solution. I was hesitate to throw errors if it doesn't exist; I'm happy to change it where it will throw if it makes sense.

Reply:
Yes.  i would validate all argument passed in.  I would expect the set of file to exactly match the set of docs in the project.  if it doesn't (because it has too many, too few, or mismatched file names) i would throw and say that the caller is doing something wrong.